src/Types/Animations/BubblePop.tsx

```diff
@@ -1,98 +1,209 @@
+            case("RECT"):
+                xpos += dim/2;
+                ypos += dim/2;
+
+                cache = this.squareCache;
+                shape = SelectedShape.Rectangle;
+                // color = getRectColor();
+                break;
+            case("POLY"):
+                cache = this.triangleCache;
+                shape = SelectedShape.Triangle;
+                // color = getTriangleColor();
+                break;
+        }
+
+        if(dim > 25) { 
+            if(shape && sketchData.figs.length + 5 <= this.NUM_FIGURES) {
+                let newFig : AnimatedFigure;
+                let numNewFigs = dim/10 > 5 ? 5 : Math.floor(dim/10);
+                numNewFigs = Math.min(numNewFigs, cache.length);
+                for(let i = 0; i < numNewFigs; ++i) {
+                    newFig = cache.shift();
+                    newFig.pos.x = xpos + ((Math.random() - 0.5) * 2)*dim/2;
+                    newFig.pos.y = ypos + ((Math.random() - 0.5) * 2)*dim/2;
+                    newFig.dim = dim * (0.3 + Math.random() * 0.3);
+
+                    sketchData.figs.push(newFig);
+                }
+            }
         }
+        let recycleFig = sketchData.figs.splice(index, 1)[0]
+        // recycleFig.color = 
+        cache.push(recycleFig);
+        fig.pop.play();
     }
+    
+    static burst(fig, sketchData, p) {
+        let index = sketchData.figs.indexOf(fig);
+
+        let type : string = fig.getShapeDescriptor().type;
+        let dim : number = fig.dim;
+        let xpos : number = fig.pos.x;
+        let ypos : number = fig.pos.y;
+
+        for(let i=0; i < 3; ++i){
+            fig.dim += 5;
+            fig.display();
+        }
+
+        if(dim > 25) {
+            let shape : SelectedShape;
+
+            switch(type) {
+                case("CIRCLE"):
+                    shape = SelectedShape.Circle;
+                    break;
+                case("RECT"):
+                    xpos += dim/2;
+                    ypos += dim/2;
+                    shape = SelectedShape.Rectangle;
+                    break;
+                case("POLY"):
+                    shape = SelectedShape.Triangle;
+                    break;
+            }
+            
+            if(shape && sketchData.figs.length + 5 <= this.NUM_FIGURES) {
+                let newFig : AnimatedFigure;
+                let numNewFigs = dim/10 > 5 ? 5 : Math.floor(dim/10);
+                for(let i = 0; i < numNewFigs; ++i) {
+                    let xOffset = ((Math.random() - 0.5) * 2)*dim/2;
+                    let yOffset = ((Math.random() - 0.5) * 2)*dim/2;
+                    newFig = newFigure(shape, xpos + xOffset, ypos + yOffset, p);
+                    newFig.dim = dim * (0.3 + Math.random() * 0.3);
+                    sketchData.figs.push(newFig);
+                }
+            }
+        }
+        sketchData.figs.splice(index, 1);
+        fig.pop.play();
+    }
+
+    static cacheFigures(sketchData, p) {
```

Comment:
I don't really get what the cache is doing, but the animation looks good so maybe lets talk about it wednesday if you think it can help make other animations more smooth.

Reply:
Previously, burst was calling newFigure() up to 5 times during a single draw loop which was causing it to stutter. With the cache, at most one newFigure() will be created in a draw loop. 

I'm not sure if it will help other animations because they create newFigures in the mousePressed() event which doesn't effect the draw loop??

Reply:
interesting, I wouldn't think that `newFigure` would be a bottleneck. Good to know though